### PR TITLE
 Basic auth

### PR DESCRIPTION
**Pull request by davisuga**
- lib update
- added needed providers
- feat:  basic user types
- feat: basic auth reducer
- feat: basic auth context
- feat: basic visual offline login
- feat: basic relay config
- feat: added createUser mutation
- feat: added generated queries
- feat(login): added query
- feat: added sign up mutation

### CODE REVIEW

File: src/RelayEnv.res [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ */
+let fetchQuery: RescriptRelay.Network.fetchFunctionPromise = (
+  operation,
+  variables,
+  _cacheConfig,
+  _uploadables,
+) => {
+  open Fetch
+  fetchWithInit(
+    "http://192.168.1.187:8080/graphql",
```
Comment (cl3t0) [inline src/RelayEnv.res]: You should not hard code an IP.

File: src/PersistReducer.res [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+let dumpState = (key, state) =>
+  switch Js.Json.stringifyAny(state) {
+  | Some(encoded) => Ok(setItem(key, encoded))
+  | None => Error("Could not encode state")
+  }
+
+let restoreState = key => getItem(key)->Promise.thenResolve(decodeJSStr)
+
+let persistReducer = (key, reducer, state, action) => {
+  let result = reducer(state, action)
```
Comment (cl3t0) [inline src/PersistReducer.res]: You can change the variable name to something more meaningful like `newState`.

File: src/navigation/Navigation.res [hunk 3]
```diff
@@ -1,25 +1,33 @@
...
-  }
-  include Stack.Make(StakeParams)
+  include MainStack
   @react.component
-  let make = (~navigation as _, ~route as _) =>
-    <Navigator>
-      <Screen name="Login" component=Login.make />
-      <Screen name="Home" component=Home.make />
+  let make = (~navigation as _, ~route as _) => {
+    let (state, _dispatch) = Context.AuthContext.use()
```
Comment (cl3t0) [inline src/navigation/Navigation.res]: You should use UseAuth instead of AuthContext here.

File: src/navigation/Navigation.res [hunk 4]
```diff
@@ -1,25 +1,33 @@
...
+
+    <Navigator screenOptions={_optionsProps => options(~headerShown=false, ())}>
+      {switch state {
+      | UserDomain.LoggedIn(_) => <Screen name="Home" component=Home.make />
+      | UserDomain.LoggedOut =>
+        <>
+          <Screen name="Login" component=Login.make />
+          <Screen name="InsertCode" component=InsertCode.make />
```
Comment (cl3t0) [inline src/navigation/Navigation.res]: Missing SignUp screen here, no?

**Review by cl3t0 — CHANGES_REQUESTED** [review]

File: src/RelayEnv.res (hunk 1 above)
Comment (davisuga) [reply src/RelayEnv.res]: Fixed

File: src/PersistReducer.res (hunk 2 above)
Comment (davisuga) [reply src/PersistReducer.res]: Fixed

File: src/navigation/Navigation.res (hunk 3 above)
Comment (davisuga) [reply src/navigation/Navigation.res]: Fixed

**Review by cl3t0 — APPROVED** [review]

**Pull request merged**